DBZ-9507: Add new article about "Adding a new table with Debezium: Best Practices and Pitfalls" (#1178)
debezium/debezium.github.io: _posts/2025-09-30-add-new-table-to-capture-list.adoc

`table.include.list`:: An optional, comma-separated list of regular expressions that match fully-qualified table identifiers for tables whose changes you want to capture. When this property is set, the connector captures changes only from the specified tables. Each identifier is of the form schemaName.tableName. By default, the connector captures changes in every non-system table in each schema whose changes are being captured.

there are also properties to apply opposite logic:

`database.exclude.list`:: An optional, comma-separated list of regular expressions that match the names of databases from which you do not want the connector to capture changes.
`schema.exclude.list`:: An optional, comma-separated list of regular expressions that match names of schemas for which you do not want to capture changes.
`table.exclude.list`:: An optional, comma-separated list of regular expressions that match fully-qualified table identifiers for tables whose changes you do not want to capture.

> the final configuration of these properties tells Debezium which tables to capture events from.

Reviewer: Here would be good just to remind that you can use either include or exclude but not both.

_How Debezium gets table schema information?_

Debezium, in order to start streaming data changes from a table, needs to know its schema so it can provide information about the table’s structure.

> Normally, Debezium learns about table schemas while streaming by reading DDL statements, parsing them, and updating an in-memory representation of each table’s schema.

Reviewer: I think this part needs an expanison. In fact there are mutliple options here
DDL - e.g. MariaDB, MySQL, Oracle
metadata message - e.g. PostgreSQL
JDBC - e.g. SQL Server, Db2

Author: If I get it correctly the JDBC approach is the DML one so that when a change arrive for a table in the capture list but not in the schema representation its schema will be captured. Isn't it?

> This information is also persisted so that it can survive a connector restart or failure.

Reviewer: That's correct only for subset of connectors, those that don't resend the metadata.

For this purpose, there are several properties prefixed with `schema.history.internal.` that manage configuration in this area.
Of particular interest is the property:

`schema.history.internal.store.only.captured.tables.ddl`:: Specifies whether the connector records schema structures from all tables in a schema or database, or only from tables that are designated for capture.

For now remember it, we will see in detail what role it has later.

Now, let’s take a step back: what happens when a connector is started for the first time?

> Debezium requires an initial snapshot for the very same reason.

Reviewer: ```suggestion
Debezium requires capuring of database schema for the very same reason.
```

Before it can start emitting events for a table, it needs to know the table structure. And even though this information can be obtained from a DDL change while streaming,
when the connector starts for the very first time it’s quite unlikely that this DDL is still present in the logs.

> This is why Debezium requires the initial snapshot, and you can configure its behavior via the `snapshot.mode` configuration.

Reviewer: ```suggestion
This is one of the reasons why Debezium requires the initial snapshot, and you can configure its behavior via the `snapshot.mode` configuration.
```

> As a side note, the initial snapshot is required for this reason, but it can also be useful for snapshotting the current database state — both data and schema — before streaming begins.

Reviewer: In general the text should be modified a bit. Initial snapshot was introduced to the exisitng data and capturing the schema is the useful by-product.

So capturing the schema is always must, initial snapshot is an extension.

If you’re not interested in capturing the existing data in the table, you can simply rely on the `no_data` snapshot mode.

> This is true for the majority of connectors, except PostgreSQL.

Reviewer: The need to capture the schema is true

date:   2025-09-30
tags: [ cdc, integration, configuration, howto ]
author: mfvitale
---

"How can I add a new table to start capturing its changes?"

This is one of the most common questions from our community.

If you ask The Ultimate Hitchhiker’s Guide to the Galaxy, the answer to everything is 42.
Sadly, in the real world, it’s not quite that simple. The real answer is: it depends.

With this article, I’d like not only to provide an answer by walking through different scenarios,
but also to explain the reasons behind it.

+++<!-- more -->+++

== Introduction
Before exploring all the possibilities, we first need to define some basic Debezium concepts.

_How does Debezium know which tables to capture?_

Debezium offers configurations to set which database/schema/table to capture via the following properties:

`database.include.list`:: An optional, comma-separated list of regular expressions that match the names of the databases from which the connector captures changes. The connector does not capture changes in any database whose name is not in database.include.list. By default, the connector captures changes in all databases.

> `schema.include.list`:: An optional, comma-separated list of regular expressions that match names of schemas for which you want to capture changes. Any schema name not included in schema.include.list is excluded from having its changes captured. By default, all non-system schemas have their changes captured.

Reviewer 2: It'd be essential to highlight somewhere in this section that not every connector supports all three of these options.

For example MySQL/MariaDB use database + table while Oracle/PostgreSQL use schema + table.

Author: Good point. This is always something that create confusion also to me. I'll add it.

There is also a related property:

`schema.history.internal.store.only.captured.databases.ddl`:: Specifies whether the connector records schema structures from all databases or only from databases that are designated for capture.

For now remember it, we will see in detail what role it has later.

Now, let’s take a step back: what happens when a connector is started for the first time?

Before emitting events for a table, Debezium needs to know the table structure.
Even though this information could be obtained from a DDL change while streaming, it’s unlikely that the DDL is still present in the logs when the connector starts for the very first time.
This is why Debezium requires capturing the database schema at first start-up.

As a side note, the initial snapshot exists primarily to capture the current data present in the database,
since not all data may be available in the transaction log when Debezium starts for the very first time.
Capturing the schema is required so that Debezium knows the table structures.
Even if you don’t want to capture the existing data, schema capture cannot be disabled.

> You can, however, skip capturing the data itself by using the no_data snapshot mode.

Reviewer 2: ```suggestion
You can, however, skip capturing the data itself by using the `no_data` snapshot mode.
```

That said, you should start to understand why the answer is “it depends.”

Let’s walk through a couple of common scenarios you may run into.

=== Scenario 1

> Let’s start with an easy scenario: you’ve set the `*.include.list` properties and ended up capturing all available tables in the database.

Reviewer 2: I think you may have intended this to read "you've not set the `*.include.list` properties" ?

Author: Well the intention here is you have set the combination of `database.include.list`, `schema.include.list`, and `table.include.list` so that you effectively capture all tables. Dunno if it's clear...

Reviewer 2: I see. In that case, then perhaps it makes sense to clarify that the same can be accomplished by leaving the include/exclude configurations unset and all tables are captured?

First of all, in this scenario, the `schema.history.internal.store.only.captured.tables.ddl` property has no impact, since all table schemas are captured anyway.

If you create a new table while Debezium is streaming, it captures the DDL and updates its internal schema representation.

> When DML occurs on that table, Debezium starts emitting change events immediately, since the table structure is already known.

Reviewer 2: ```suggestion
When DML changes occur on that table, Debezium starts emitting change events immediately, since the table structure is already known.
```

. Add the table to the capture list.
. Restart the connector.

Basically, this puts Debezium in a “first run” mode just for the schema. No existing data will be re-snapshotted.

NOTE: We have implemented an experimental feature for Oracle that automatically registers the schema for a table in-flight when a DML occurs.
This means that if Debezium observes an insert, update, or delete event for a table whose schema is not yet captured and the table matches the include list,
it will attempt to register the current schema and process the event using that schema.

=== Wait! What about MongoDB?

Due to its schemaless nature, MongoDB doesn’t require you to predefine a schema like relational databases do.
So if you want to add a new collection to be captured, you just need to:

. Ensure that the configured `capture.scope` is set to include the desired collection.
. Edit the `collection.include.list` to add the new collection.

=== Conclusion

We have seen how to add a new table to the captured list in different scenarios, and also explored the reasons behind the required steps.
I hope this improves your Debezium knowledge and helps you manage it more confidently.

> Here at Debezium, we are always researching ways to improve the project and its usability, and this article is part of that effort.

Reviewer 2: ```suggestion
The Debezium team is continually seeking ways to enhance the project and its usability, and this article is part of that ongoing effort.
```
